SPECS-SIGNED/kernel-signed-aarch64/kernel-signed-aarch64.spec

```diff
@@ -3,63 +3,103 @@
+# This spec purpose is to take an input kernel rpm and input secure-boot-signed
+# kernel binary from the same build and generate a new "kernel" rpm with the
+# signed kernel binary + all of the other original kernel files, triggers,
+# scriptlets, requires, provides, etc.
 #
-# To populate these sources:
+# We need to ensure the kernel modules and kernel binary used are from the exact
+# same build because at build time the kernel modules are signed with an
+# ephemeral key that the kernel enrolls in its keyring. We enforce kernel
+# module signature checking when we enable security features like kernel
+# lockdown so our kernel can only load those specific kernel modules at runtime.
+#
+# Additionally, to complete the UEFI Secure Boot chain, we must PE-sign the
+# kernel binary. Ideally we would enable secure-boot signing tools like pesign
+# or sbsign to be callable from inside the rpmbuild environment, that way we can
+# secure-boot sign the kernel binary during the kernel's rpmbuild. It is best
+# practice to sign as soon as possible. However there are issues getting that
+# secure boot signing infrastructure in place today. Hence we sign the
+# resulting kernel binary and "repackage" the kernel RPM (something rpm itself
+# actively tries to make sure you never do...generally for good reasons).
+#
+# To achive this repackaging, this spec creates a new subpackage named
+# "kernel". To retain all of the initial kernel package behaviors, we make sure
+# the subpackage has the same requires, provides, triggers, post steps, and
+# files as the original kernel package.
+#
+# This specific repackaging implementation leaves room for us to enable the
+# more ideal secure-boot signing flow in the future without introducing any
+# sort of breaking change or new packaging. Users still install a "kernel"
+# package like they normally would.
+#
+# Maintenance Notes:
+# - This spec's "version" and "release" must reflect the unsigned version that
+# was signed. An important consequence is that when making a change to this
+# spec or the normal kernel spec, the other spec's version version/release must
+# be increased to keep the two versions consistent.
+#
+# - Make sure the kernel subpackage's Requires, Provides, triggers, post/postun
+# scriptlets, and files match the normal kernel spec's. The kernel subpackage
+# should contain the same content as the input kernel package but replace the
+# kernel binary with our signed kernel binary. Since all the requires, provides,
+# etc are the same, this new kernel package can be a direct replacement for the
+# normal kernel package and RPM will resolve packages with kernel dependencies
+# correctly.
+#
+# To populate the input sources:
 #   1. Build the unsigned packages as normal
 #   2. Sign the desired binary
 #   3. Place the unsigned package and signed binary in this spec's folder
 #   4. Build this spec
 Source0:        kernel-%{version}-%{release}.aarch64.rpm
```

Comment:
I don't understand why we use source file instead of depending on a RPM package. This makes me feel a bit uneasy, especially when considering that different sources might end up having the same name.

Reply:
When you say "depending on an RPM package", I presume you mean using it as a BuildRequires? BuildRequires will install the RPM components into the chroot environment's folders. Since my goal is to remake the exact same kernel RPM package just with the signed vmlinuz inside, it seemed safer/better to extract the contents of the kernel RPM into BUILD instead of try to fish out each file from the chroot's system folders, potentially including or leaving behind files.

Reply:
I don't think the list of the files should be a problem. At the very least, you could query the existing RPM database for the package's files, print to file and use `%files -f list_of_files.txt`. This way we don't depend on some external source providing us with the rpm file, but use the same mechanism for all deps.